thavalon-server/src/game/mod.rs

```diff
@@ -0,0 +1,247 @@
+    roles: HashMap<Role, PlayerId>,
+    good_players: Vec<PlayerId>,
+    evil_players: Vec<PlayerId>,
+}
+
+impl Players {
+    fn new() -> Players {
+        Players {
+            players: HashMap::new(),
+            roles: HashMap::new(),
+            good_players: Vec::new(),
+            evil_players: Vec::new(),
+        }
+    }
+
+    fn add_player(&mut self, player: Player) {
+        self.roles.insert(player.role, player.id);
+        if player.role.is_good() {
+            self.good_players.push(player.id);
+        } else {
+            self.evil_players.push(player.id);
+        }
+        self.players.insert(player.id, player);
+    }
+
+    fn by_role(&self, role: Role) -> Option<&Player> {
+        self.roles.get(&role).map(|id| &self[*id])
+    }
+
+    fn has_role(&self, role: Role) -> bool {
+        self.roles.contains_key(&role)
+    }
+
+    fn good_players(&self) -> &[PlayerId] {
+        self.good_players.as_slice()
+    }
+
+    fn evil_players(&self) -> &[PlayerId] {
+        self.evil_players.as_slice()
+    }
+
+    fn iter(&self) -> impl Iterator<Item = &Player> {
+        self.players.values()
+    }
+
+    fn len(&self) -> usize {
+        self.players.len()
+    }
+}
+
+impl Index<PlayerId> for Players {
+    type Output = Player;
+
+    fn index(&self, player_id: usize) -> &Self::Output {
+        &self.players[&player_id]
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub enum Card {
+    Success,
+    Fail,
+    Reverse,
+}
+
+impl fmt::Display for Card {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str(match self {
+            Card::Success => "Success",
+            Card::Fail => "Fail",
+            Card::Reverse => "Reverse",
+        })
+    }
+}
+
+/// Game rules determined by the number of players
+#[derive(Debug, Clone)]
+pub struct GameSpec {
+    /// Number of players in the game
+    pub players: u8,
+    /// The number of players on each mission
+    pub mission_sizes: [usize; 5],
+    /// Allowed good roles in the game
+    pub good_roles: &'static [Role],
+    /// Allowed evil roles in the game
+    pub evil_roles: &'static [Role],
+    /// The number of players on the good team
+    pub good_players: u8,
+    /// True if mission 4 requires at least two failures
+    double_fail_mission_four: bool,
+}
+
+impl GameSpec {
+    pub fn for_players(players: usize) -> &'static GameSpec {
+        match players {
+            5 => &FIVE_PLAYER,
+            _ => panic!("{}-player games not supported", players),
+        }
+    }
+
+    pub fn mission_size(&self, mission: MissionNumber) -> usize {
+        self.mission_sizes[mission as usize - 1]
+    }
+
+    pub fn good_players(&self) -> usize {
+        self.good_players as usize
+    }
+
+    pub fn evil_players(&self) -> usize {
+        (self.players - self.good_players) as usize
+    }
+
+    /// The number of proposals in a round (after the first, which only has two)
```

Comment:
So we're not supporting the version with infinite proposals until only force currently? Not sure how much that'll change things, so might be easy to change in the future

Reply:
Honestly I forgot about that version - I think I only played it once. We can switch to it by modifying the loop in `run_game` pretty easily though, I think